drivers/sdhc/emmc_host.c

```diff
@@ -0,0 +1,1337 @@
+		if (cmd->opcode == SD_SEND_STATUS) {
+			k_msleep(10);
+		}
+	}
+
+	return ret;
+}
+
+static int emmc_set_io(const struct device *dev, struct sdhc_io *ios)
+{
+	struct emmc_data *emmc = dev->data;
+	volatile struct emmc_reg *regs = emmc->regs;
+	struct sdhc_io *host_io = &emmc->host_io;
+	int ret;
+
+	LOG_DBG("emmc I/O: DW %d, Clk %d Hz, card power state %s, voltage %s", ios->bus_width,
+		ios->clock, ios->power_mode == SDHC_POWER_ON ? "ON" : "OFF",
+		ios->signal_voltage == SD_VOL_1_8_V ? "1.8V" : "3.3V");
+
+	if (ios->clock && (ios->clock > emmc->props.f_max || ios->clock < emmc->props.f_min)) {
+		LOG_ERR("Invalid argument for clock freq: %d Support max:%d and Min:%d", ios->clock,
+			emmc->props.f_max, emmc->props.f_min);
+		return -EINVAL;
+	}
+
+	/* Set HC clock */
+	if (host_io->clock != ios->clock) {
+		LOG_DBG("Clock: %d", host_io->clock);
+		if (ios->clock != 0) {
+			/* Enable clock */
+			LOG_DBG("CLOCK: %d", ios->clock);
+			if (!emmc_clock_set(emmc, ios->clock)) {
+				return -ENOTSUP;
+			}
+		}
+		host_io->clock = ios->clock;
+	}
+
+	/* Set data width */
+	if (host_io->bus_width != ios->bus_width) {
+		LOG_DBG("bus_width: %d", host_io->bus_width);
+		SET_BITS(regs->host_ctrl1, EMMC_HOST_CTRL1_EXT_DAT_WIDTH_LOC,
+			 EMMC_HOST_CTRL1_EXT_DAT_WIDTH_MASK,
+			 ios->bus_width == SDHC_BUS_WIDTH8BIT ? 1 : 0);
+
+		regs->host_ctrl1 |= ios->bus_width << EMMC_HOST_CTRL1_DAT_WIDTH_LOC;
+		host_io->bus_width = ios->bus_width;
+	}
+
+	/* Set HC signal voltage */
+	if (ios->signal_voltage != host_io->signal_voltage) {
+		LOG_DBG("signal_voltage: %d", ios->signal_voltage);
+		ret = emmc_set_voltage(emmc, ios->signal_voltage);
+		if (ret) {
+			LOG_ERR("Set signal volatge failed failed:%d", ret);
+			return ret;
+		}
+		host_io->signal_voltage = ios->signal_voltage;
+		k_msleep(10u);
+	}
+
+	/* Set card power */
+	if (host_io->power_mode != ios->power_mode) {
+		LOG_DBG("power_mode: %d", ios->power_mode);
+
+		ret = emmc_set_power(emmc, ios->power_mode);
+		if (ret) {
+			LOG_ERR("Set Bus Width failed:%d", ret);
+			return ret;
+		}
+		host_io->power_mode = ios->power_mode;
+	}
+
+	/* Set I/O timing */
+	if (host_io->timing != ios->timing) {
+		LOG_DBG("timing: %d", ios->timing);
+
+		ret = set_timing(emmc, ios->timing);
+		if (ret) {
+			LOG_ERR("Set timing failed:%d", ret);
+			return ret;
+		}
+		host_io->timing = ios->timing;
+	}
+
+	return ret;
+}
+
+static int emmc_get_card_present(const struct device *dev)
+{
+	struct emmc_data *emmc = dev->data;
+	volatile struct emmc_reg *regs = emmc->regs;
+
+	LOG_DBG("");
+
+	emmc->card_present = (bool)((regs->present_state >> 16u) & 1u);
+
+	if (!emmc->card_present) {
+		LOG_ERR("No MMC device detected");
+	}
+
+	return ((int)emmc->card_present);
+}
+
+static int emmc_execute_tuning(const struct device *dev)
+{
+#if defined(CONFIG_EMMC_HOST_TUNING)
+	struct emmc_data *emmc = dev->data;
+	volatile struct emmc_reg *regs = emmc->regs;
+
+	LOG_ERR("Tuning starting...");
```

Comment:
Why use `LOG_ERR` here? does the user need this log any time they enable basic SDMMC logging?